presto-native-execution/presto_cpp/main/expression/RowExpressionOptimizer.cpp

```diff
@@ -0,0 +1,709 @@
+ * limitations under the License.
+ */
+#include "presto_cpp/main/expression/RowExpressionOptimizer.h"
+#include <proxygen/httpserver/ResponseBuilder.h>
+#include "presto_cpp/main/common/Utils.h"
+#include "velox/common/encode/Base64.h"
+#include "velox/exec/ExchangeQueue.h"
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/ExprCompiler.h"
+#include "velox/expression/FieldReference.h"
+#include "velox/expression/VectorFunction.h"
+
+using namespace facebook::presto;
+using namespace facebook::velox;
+
+namespace facebook::presto::expression {
+
+namespace {
+
+protocol::TypeSignature getTypeSignature(const TypePtr& type) {
+  std::string typeSignature;
+  if (type->parameters().empty()) {
+    typeSignature = type->toString();
+    boost::algorithm::to_lower(typeSignature);
+  } else if (type->isDecimal()) {
+    typeSignature = type->toString();
+  } else {
+    std::vector<TypePtr> childTypes;
+    if (type->isRow()) {
+      typeSignature = "row(";
+      childTypes = asRowType(type)->children();
+    } else if (type->isArray()) {
+      typeSignature = "array(";
+      childTypes = type->asArray().children();
+    } else if (type->isMap()) {
+      typeSignature = "map(";
+      const auto mapType = type->asMap();
+      childTypes = {mapType.keyType(), mapType.valueType()};
+    } else {
+      VELOX_USER_FAIL("Invalid type {}", type->toString());
+    }
+
+    if (!childTypes.empty()) {
+      auto numChildren = childTypes.size();
+      for (auto i = 0; i < numChildren - 1; i++) {
+        typeSignature += fmt::format("{},", getTypeSignature(childTypes[i]));
+      }
+      typeSignature += getTypeSignature(childTypes[numChildren - 1]);
+    }
+    typeSignature += ")";
+  }
+
+  return typeSignature;
+}
+
+json toVariableReferenceExpression(
+    const std::shared_ptr<const exec::FieldReference>& fieldReference) {
+  protocol::VariableReferenceExpression vexpr;
+  vexpr.name = fieldReference->name();
+  vexpr._type = "variable";
+  vexpr.type = getTypeSignature(fieldReference->type());
+  json res;
+  protocol::to_json(res, vexpr);
+
+  return res;
+}
+
+bool isPrestoSpecialForm(const std::string& name) {
+  static const std::unordered_set<std::string> kPrestoSpecialForms = {
+      "and",
+      "coalesce",
+      "if",
+      "in",
+      "is_null",
+      "or",
+      "switch",
+      "when",
+      "null_if"};
+  return kPrestoSpecialForms.count(name) != 0;
+}
+
+json::array_t getInputExpressions(
+    const std::vector<std::unique_ptr<folly::IOBuf>>& body) {
+  std::ostringstream oss;
+  for (auto& buf : body) {
+    oss << std::string((const char*)buf->data(), buf->length());
+  }
+  return json::parse(oss.str());
+}
+
+template <TypeKind KIND>
+std::shared_ptr<exec::ConstantExpr> getConstantExpr(
+    const TypePtr& type,
+    const DecodedVector& decoded,
+    memory::MemoryPool* pool) {
+  std::shared_ptr<exec::ConstantExpr> constExpr = nullptr;
```

Comment:
There isn't a need to declare this variable. Since the If part fails, the else part is not required and the make_shared line at 117 can directly return.